### ISSUE TITLE
About the bos token in Gemma4 finetune notebook

### ISSUE THREAD

**Issue by omegayal1234** [unslothai/notebooks#251]
In the notebook for [Gemma4 (31B) Conversational](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Gemma4_(31B)-Text.ipynb), the leading `<bos>` token is removed using `removeprefix('<bos>') ` when formatting the prompt, and is said that it will be added back before training.
However, while verifying whether the `<bos>` token is appeared only once in the training sample via `tokenizer.decode(trainer.train_dataset[100]["input_ids"])`, the result clearly shows no `<bos>` at all. In addition, I've also check that the result of  `tokenizer.tokenizer.add_bos_token` is `False`.
Does this means that the `<bos>` token is truly disappeared during the training process?
Do we need to add it back to ensure that the prompt format is correct as expected?

**Comment by Ki-Seki** [comment]
Hi. I’ve also encountered this issue. Did you manage to resolve it? I suspect the Unsloth team might have made a mistake here.


## We cannot see the bos in the printed training data example

<img width="2879" height="1768" alt="Image" src="https://github.com/user-attachments/assets/18c64730-3c21-4d1b-8b6a-34fd96b14733" />

## Training loss indicates missing `<bos>` cause model training collapse

### Training Loss of GPT (loss starting from about 2~)

<img width="2879" height="1766" alt="Image" src="https://github.com/user-attachments/assets/e7aa2af6-3eba-47e0-919d-f1904d5952ef" />

### Training Loss of Qwen (loss starting from about 1~)

<img width="2878" height="1751" alt="Image" src="https://github.com/user-attachments/assets/dd54f8f4-02c2-4096-a139-8289ea6a4f06" />

### Training Loss of Gemma4 (loss starting from about 12~ !!!!🤯)

<img width="2879" height="1760" alt="Image" src="https://github.com/user-attachments/assets/e7d1d9ea-aa10-4632-9596-eecfdd02b112" />

### This is also the same in my experiments

<img width="2155" height="1344" alt="Image" src="https://github.com/user-attachments/assets/55219ee9-fa0e-475a-8b6e-dc5392060573" />

**Comment by Ki-Seki** [comment]
@danielhanchen
